### ISSUE TITLE
False Positives

### ISSUE THREAD

**Issue by cassin-edwin** [huggingface/setfit#352]
I had built a model using a muti-label dataset. But I see that I am getting so many False Positive outputs during inference.

For eg:

FIRST NOTICE OF LOSS SENT TO AGENT'S CUSTOMER ACTIVITY    ---> This is predicted as 'Total Loss' (Total Loss is one of my labels given fed through the dataset). 

I see that there is a word 'Loss' present in the dataset but it is not supposed to be predicted as 'Total Loss'.

There are so many absurd outputs as well.

Here is the pre-trained model which I am using for fine-tuning :
Hyper-parameters : num_iterations = 30, batch_size = 16, num_epochs = 1

What went wrong?

**Comment by tomaarsen** [comment]
Those are some reasonable hyperparameters. What is the Sentence Transformers that you're using? `"sentence-transformers/paraphrase-mpnet-base-v2"`?
Via `compute_metrics` you might be able to write an evaluation method to get an idea of the performance. I'm not sure how else I can help.

- Tom Aarsen

**Comment by tomaarsen** [comment]
You can use the `metrics` parameter of the Trainer:
https://github.com/huggingface/setfit/blob/d9aff37af91f200ed640839d1aebe4c9a96e9563/src/setfit/trainer.py#L43-L45

You can use a string here, like "f1", and it'll take it from these spaces: https://huggingface.co/evaluate-metric, so for F1 it takes https://huggingface.co/spaces/evaluate-metric/f1/blob/main/f1.py.

However, I think people have reported some issues with multi-label datasets. In that case, you can also provide a function like:
```python
import evaluate

def compute_metrics(y_pred, y_test) -> Dict[str, float]:
    f1_metric = evaluate.load("f1")
    # or "macro" or "weighted" or None as the average
    return f1_metric.compute(references=y_test, predictions=y_pred, average="micro")

trainer = SetFitTrainer(
    ...,
    metrics=compute_metrics,
    ...
)

...

trainer.evaluate()
```

I think that accuracy might not be too awful depending on the number of classes and the amount of training data. My experience is that extra training data always does wonders for the performance.

- Tom Aarsen

**Comment by tomaarsen** [comment]
I think this is ready to be closed.

**Comment by tomaarsen** [comment]
There are some additional steps that you could take to try and get better performance:
* Try a different underlying sentence-transformer, i.e. from this list: https://huggingface.co/sentence-transformers (or perhaps even common models like `bert-base-uncased`, I've seen one example where a user experienced better performance using some RoBERTa-inspired model as the sentence-transformer)
* Perform hyperparameter tuning, e.g. like in https://github.com/huggingface/setfit/blob/main/notebooks/text-classification_hyperparameter-search.ipynb
* Label more data. This one tends to be very important. You can estimate whether your model will benefit from additional data by retraining the model with e.g. 20% less data. If the performance does not drop much, then additional data is unlikely to improve much. However, if it drops a lot, then the small bit of data was quite important, and more additional data may then also be very important.

With an accuracy of 59%, it is not too unexpected that the model sometimes makes mistakes and gives false positives.

- Tom Aarsen

**Comment by tomaarsen** [comment]
> But that didn't work because we can specify only the metric name. 

You can also provide a function, like:
```python
def compute_metrics(y_pred, y_test) -> Dict[str, float]:
    f1_metric = evaluate.load("f1")
    # or "macro" or "weighted" or None as the average
    return f1_metric.compute(references=y_test, predictions=y_pred, average="micro")

trainer = SetFitTrainer(
    ...,
    metrics=compute_metrics,
    ...
)
trainer.evaluate()
```

And since SetFit v0.7.0 you can use the new `metric_kwargs` parameter on the `SetFitTrainer`, which was developed specifically for this problem. You can see an example here: https://github.com/huggingface/setfit/blob/92e966b38d295705a1cb082bf6dbec3ca5b89b76/tests/test_trainer.py#L469-L487

In short, add `metric_kwargs={"average": "micro"}` to the Trainer.
